### ISSUE TITLE
Errorbar semilogy switch to log y removed error bars when crosses into negative

### ISSUE THREAD

**Issue by PaulKuin** [matplotlib/matplotlib#9149]
### Bug report

**Bug summary**

When plotting error bars on data points, where the data is positive but the bar with the data-error is entering negative values, switching to log Y axes using  semilogy() means that some data points are plotted without error bars at all (both y-err bars disappear).

**Code for reproduction**

create a file example.dat:
   x               dx                 y              dy
1.046850 0.0029168 0.042551 0.02420 
1.506679 0.0030152 0.027573 0.02252 
3.028597 0.1959554 0.021218 0.00974 
3.656348 0.1034119 0.002788 0.01059 
4.949535 0.4693247 0.006276 0.00969 

from astropy.io import ascii
from pylab import figure, errorbar,semilogy
t = ascii.read('example.dat')
figure()
errorbar(t['x'],t['y'],xerr=t['dx'],yerr=t['dy'],fmt='s')

This shows a linear plot with errorbars fine. The third data point error bar extends into negative territory.

semilogy()

Now the plot has a log y axis, but the error bar of the third point is gone.

```python
# Paste your code here
#
#
```
from astropy.io import ascii
from pylab import figure, errorbar,semilogy
t = ascii.read('example.dat')
figure()
errorbar(t['x'],t['y'],xerr=t['dx'],yerr=t['dy'],fmt='s')

#This shows a linear plot with errorbars fine. The third data point error bar extends into negative territory.

semilogy()

#Now the plot has a log y axis, but the error bar of the third point is gone.
**Actual outcome**

<img width="642" alt="screen shot 2017-09-01 at 15 43 50" src="https://user-images.githubusercontent.com/5159688/29974969-96166f08-8f2c-11e7-860d-b0359bc89a5d.png">
<img width="638" alt="screen shot 2017-09-01 at 15 44 05" src="https://user-images.githubusercontent.com/5159688/29974980-9b15bee6-8f2c-11e7-9da1-0148a620886b.png">

```
# If applicable, paste the console output here
#
#
```

**Expected outcome**

An errorbar in the log plot, surely the positive bar, and the negative going a few (3 or 5) decades down (perhaps with an arrow ending).

**Matplotlib version**

  * Operating System: OSX
  * Matplotlib Version: 2.0.2
  * Python Version:2.7
  * Jupyter Version (if applicable):
  * Other Libraries: astropy

anaconda astroconda

**Comment by dstansby** [comment]
Yeah, we change the behaviour of this recently to fix some other issues (#8836). There is a note in the docstring of `errorbar` that what you're describing is expected behaviour, but I do agree that we should find some way to put the errorbars back in and have them go right down.

**Comment by tacaswell** [comment]
I think the issue here is that exactly what should be done with the negative is not clear, I think there are at least 4 options:

 - mask the bottom point and drop the whole errorbar (it is implemented as a single line)
 - mask the bottom point, but still draw the top half (not done, but would be possible without too much trouble)
 - clip to some 'small' number (now picking that small number is it's own issue and now we have to be careful on auto-scaling to not include that in the limits) (and this is implemented via a kwarg flag)
   - should we clip to `0+epsilon` or `val * epsilon`? (I think we go with the first option now, but would have to check the code)
 - replace the bottom one with an arrow, but this also has to be clipped so the difference is still positive so we have to pick that number 


See https://matplotlib.org/examples/statistics/errorbar_limits.html for all the ways to have finer grain control over how limits are drawn (if you want to use log then negative values are probably non-physical and you really just have no lower bound for the value (ex neutrino mass)) so it might be worth doing the extra work to identify that.

Also see https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.set_yscale.html , if you use `ax.set_yscale('log', nonposy='clip')` it should clip the error bars to some very small number.
